runtime/server/api.go

```diff
@@ -78,3 +84,204 @@ func (s *Server) apiHandler(w http.ResponseWriter, req *http.Request) error {
 
 	return nil
 }
+
+func (s *Server) combinedOpenAPISpec(w http.ResponseWriter, req *http.Request) error {
+	// Parse path parameters
+	ctx := req.Context()
+	instanceID := req.PathValue("instance_id")
+
+	// Add observability attributes
+	observability.AddRequestAttributes(ctx)
+	s.addInstanceRequestAttributes(ctx, instanceID)
+
+	// Only GET request is allowed
+	if req.Method != http.MethodGet {
+		return httputil.Error(http.StatusMethodNotAllowed, fmt.Errorf("GET only"))
+	}
+
+	// Check if user has access to query for API data
+	if !auth.GetClaims(ctx).CanInstance(instanceID, auth.ReadAPI) {
+		return httputil.Errorf(http.StatusForbidden, "does not have access to custom APIs")
+	}
+
+	apis := make(map[string]*runtimev1.API)
+
+	// Get all built-in APIs
+	for name, api := range runtime.BuiltinAPIs {
+		apis[name] = api
+	}
+
+	// Get all custom APIs
+	ctrl, err := s.runtime.Controller(ctx, instanceID)
+	if err != nil {
+		return httputil.Error(http.StatusBadRequest, err)
+	}
+
+	list, err := ctrl.List(ctx, runtime.ResourceKindAPI, "", false)
+	if err != nil {
+		return err
+	}
+
+	for _, res := range list {
+		apis[res.Meta.Name.Name] = res.GetApi()
+	}
+
+	// Generate the OpenAPI spec
+	combinedAPISpec, err := s.generateOpenAPISpec(instanceID, apis)
+	if err != nil {
+		return httputil.Error(http.StatusInternalServerError, err)
+	}
+
+	// Check accepted format of the OpenAPI spec: JSON or YAML
+	accept := req.Header.Get("Accept")
+
+	// YAML
+	if accept == "application/yaml" || accept == "application/x-yaml" {
+		data, err := yaml.Marshal(combinedAPISpec)
+		if err != nil {
+			return httputil.Error(http.StatusInternalServerError, err)
+		}
+
+		w.Header().Set("Content-Type", "application/yaml")
+		_, err = w.Write(data)
+		if err != nil {
+			return httputil.Error(http.StatusInternalServerError, err)
+		}
+
+		return nil
+	}
+
+	// JSON
+	data, err := combinedAPISpec.MarshalJSON()
+	if err != nil {
+		return httputil.Error(http.StatusInternalServerError, err)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	_, err = w.Write(data)
+	if err != nil {
+		return httputil.Error(http.StatusInternalServerError, err)
+	}
+
+	return nil
+}
+
+func (s *Server) generateOpenAPISpec(instanceID string, apis map[string]*runtimev1.API) (*spec.Swagger, error) {
+	baseSpec := &spec.Swagger{
+		SwaggerProps: spec.SwaggerProps{
+			Swagger:  "2.0",
```

Comment:
Any reason for using v2 instead of v3?

Reply:
Majorly, to align with project swagger specs and also there is no support of v3 in [go-openapi/spec](https://github.com/go-openapi/spec?tab=readme-ov-file#faq)
There is [go-openapi/spec3](https://github.com/go-openapi/spec3) but it is incomplete

Reply:
I think we should go for OpenAPI 3 if possible because it has some useful new features and has been out for a very long time. These libraries might be worth looking into:
- https://github.com/getkin/kin-openapi (used by https://github.com/oapi-codegen/oapi-codegen)
- https://github.com/ogen-go/ogen/tree/main/openapi (subpackage of https://github.com/ogen-go/ogen/tree/main)

Reply:
Migrated to kin-openapi